### DISCUSSION TITLE
How to view HTML with embedded images

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3209] [Development]
Here's an email, where the HTML part uses embedded images.

```
1 <no description>      [multipart/alternative]
2 ├─><no description>              [text/plain]
3 └─><no description>               [text/html]
4 images/cat.gif                    [image/gif]
5 images/dog.gif                    [image/gif]
6 images/mouse.gif                  [image/gif]
```

If the user wanted to view the HTML correctly, how would they do it?

---

Each image has a unique `Content-ID:` of the form `<alphanum@alphanum>`
Any HTML references to the images are of the form `cid:CONTENT_ID` (without the `<>`s)
e.g.  `<img src="cid:1234@abcd">`

**Comment by flatcap** [comment]
Whatever the solution, the common parts will be:
- Save the HTML
- Save all the corresponding images
- Transform the HTML -- turn the `cid:1234@abcd` into the correct filenames

**Comment by gahr** [comment]
I'd implement #3200 to save all MIME parts into a common directory, and leverage an external script to do the substitutions inside the html.

**Comment by flatcap** [comment]
If we were to code this into NeoMutt, then we could add a function `<save-html-bundle>`

**Advantages** for code:
- It can work out which MIME parts **need** to be saved
- Save them with their preferred names
- It can transform the HTML using a table of cid:filename lookups
- Viewing `text/html` using mailcap can transparently save the images

**Disadvantages**:
- More code -- It may be possible to script it
- The HTML transformation would be very simple: search for quoted cid strings that match our list

This **is** just one use-case, but it's increasing common, so we want to offer the users a trivial way to get it working.

**Comment by dcpurton** [comment]
> Each image has a unique `Content-ID:` of the form `<alphanum@alphanum>` Any HTML references to the images are of the form `cid:CONTENT_ID` (without the `<>`s) e.g. `<img src="cid:1234@abcd">`

This form of `Content-ID` is not guaranteed.

e.g., Gmail uses:

```
X-Attachment-Id: ii_kyjeu0wt0
Content-ID: <ii_kyjeu0wt0>
```

Outlook uses:

```
Content-ID: <bcd6623f-e3b6-4112-af23-10dffaa6a6d2>
```

Though Thunderbird does add the domain:

```
Content-Id: <part1.V31yIz0u.0p1E8jMG@example.com>
```

**Comment by dcpurton** [comment]
I think it's best just to seamlessly hook in to `<view-attach>` and `<view-mailcap>`. Mostly, the attachment will not be in a `multipart/related` group, so there will be no difference. If the file is in a `multipart/related` group then having it modified to reference other attachments in the group is probably unlikely even to be noticed by the user.

`<save-entry>` and `<pipe-entry>` could continue to operate on the attachment unmodified.

It makes sense to me to run the command while the cursor is on the HTML part.

Were this the case then the first thing to add is a function in `attach/lib.c` which determines whether an attachment is in a group of a particular kind. Something like:

```C
/**
 * attach_body_find_ancestor - Search for a body's ancestor with specified subtype
 * @param[in]  start         Body to start search from
 * @param[in]  body          Body to find ancestor of
 * @param[in]  subtype       Ancestor MIME subtype
 * @param[out] body_ancestor Body ancestor of specified subtype if found
 * @retval     true          Ancestor of specified subtype found
 * @retval     false         Ancestor of specified subtype not found
 */
bool attach_body_find_ancestor(struct Body *start, struct Body *body,
                               char *subtype, struct Body **body_ancestor)
{
  ...
}
```

This would just involve repeated calls to `attach_body_parent()` until a parent of the right `subtype` is found or no parent is found.

Once you've got the body ancestor, then you need another function which writes out the selected attachment along with all the attachments in the group which have a `Content-ID`. (It's not necessary to write out files without a `Content-ID`.)

As files are written out you could keep track of the mapping between filename and `Content-ID`. The selected file would be written out last, filtering for any `cid:` and replacing with the mapped filename.

This function (or functions) would (I think) go in `mutt/mutt_attach.c`, since `mutt_view_attachment()` is here.

Finally, you view the file using the same logic as `<view-attach>` or `<view-mailcap>` from within `mutt_view_attachment()`.

Easy, right? :grimacing:

**Comment by dcpurton** [comment]
See #3383 for possible implementation hooking into `<view-attach>`, `<view-mailcap>` and `<view-pager>`.

`<save-entry>` and `<pipe-entry>` operate on the attachment unaltered.

**Comment by aliceinwire** [comment]
I'm doing this with ueberzug on my own configuration
```
#!/bin/bash
[ -z "$5" ] && echo "Usage: $0 <image> <x> <y> <max height> <max width>" && exit

source "`ueberzug library`"

ImageLayer 0< <(
    ImageLayer::add [identifier]="example0" [x]="$2" [y]="$3" [max_width]="$5" [max_height]="$4" [path]="$1";
    read -rs;
)

```
and mailcap 
`image/*                   ; /usr/local/bin/show %s 0 0 50 50  ; copiousoutput;`

would be nice to see html images with ueberzug in the mail body
